.github/workflows/tests.yml

```diff
@@ -0,0 +1,289 @@
+
+
+  lint-sdist:
+    if: ${{ contains(github.ref, 'release-') }}
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - uses: actions/setup-python@v2
+      with:
+        python-version: '3.x'
+    - run: pip install wheel
+    - run: python setup.py sdist bdist_wheel
+    - uses: actions/upload-artifact@v2
+      with:
+        name: distributions
+        path: dist/*
+
+
+  # Dummy step to gate other tests on without repeating the whole list
+  linting-done:
+    if: ${{ always() }} # Run this even if prior jobs were skipped
+    needs: [lint, lint-crlf, lint-newsfile, lint-sdist]
+    runs-on: ubuntu-latest
+    steps:
+      - run: true
+
+  trial:
+    if: ${{ !failure() }}  # Allow previous steps to be skipped, but not fail
+    needs: linting-done
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ['3.5', '3.6', '3.7', '3.8', '3.9']
+        database: ['sqlite']
+        include:
+          # Newest Python without optional deps
+          - python-version: '3.9'
+            toxenv: 'py-noextras,combine'
+
+          # Oldest Python with PostgreSQL
+          - python-version: '3.5'
+            database: 'postgres'
+            postgres-version: 9.5
+
+          # Newest Python with PostgreSQL
+          - python-version: '3.9'
+            database: 'postgres'
+            postgres-version: 13
+
+    steps:
+    - uses: actions/checkout@v2
+    - run: sudo apt-get -qq install xmlsec1
+    - name: Set up PostgreSQL ${{ matrix.postgres-version }}
+      if: ${{ matrix.postgres-version }}
+      run: |
+        docker run -d -p 5432:5432 \
+          -e POSTGRES_PASSWORD=postgres \
+          -e POSTGRES_INITDB_ARGS="--lc-collate C --lc-ctype C --encoding UTF8" \
+          postgres:${{ matrix.postgres-version }}
+    - uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python-version }}
+    - run: pip install tox
+    - name: Await PostgreSQL
+      if: ${{ matrix.postgres-version }}
+      timeout-minutes: 2
+      run: until pg_isready -h localhost; do sleep 1; done
+    - run: tox -e py,combine
+      env:
+        TRIAL_FLAGS: "--jobs=2"
+        SYNAPSE_POSTGRES: ${{ matrix.database == 'postgres' || '' }}
+        SYNAPSE_POSTGRES_HOST: localhost
+        SYNAPSE_POSTGRES_USER: postgres
+        SYNAPSE_POSTGRES_PASSWORD: postgres
+
+
+  trial-olddeps:
+    if: ${{ !failure() }}  # Allow previous steps to be skipped, but not fail
+    needs: linting-done
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - name: Test with old deps
+      uses: docker://ubuntu:xenial  # For old python and sqlite
+      with:
+        workdir: /github/workspace
+        entrypoint: .buildkite/scripts/test_old_deps.sh
+      env:
+        TRIAL_FLAGS: "--jobs=2"
+
+
+  trial-pypy:
+    # Very slow; only run if the branch name includes 'pypy'
+    if: ${{ contains(github.ref, 'pypy') && !failure() }}
```

Comment:
Could it be made that it also runs if the PR/commit name has "pypy"? (preferably PR name)

Reply:
It could, but I think the expression would be quite a bit longer?

I think it'd be something like `github.event == 'pull_request' && (contains(github.event.pull_request.title, 'pypy') || contains(github.ref, 'pypy') ) && !failure())`?

Reply:
Looks good, and I don't think the expression being longer is a problem, most people wouldn't see this unless they were editing this specific part of the pipeline.

Adding some comments to deconstruct what is happening could help, tho.